File: Nikhil/python/lab_agent.py
# lab_agent.py
from __future__ import annotations

import logging
from typing import Sequence

from executor_module import ExecutorModule
from execution_context import ExecutionContext
from react_agent import Step
from steps import Status
from tool import Tool, ToolCall, ToolCallResult
from toolable_react_agent import ToolableReactAgent

logger = logging.getLogger(__name__)


class LabAgent(ToolableReactAgent):
    """
    Python translation of `com.infosys.small.pnbc.LabAgent`.

    A *ReAct* agent able to execute commands in a **simulated** environment
    through an :class:`ExecutionContext`.  It can also be exposed as a *tool*
    for use by other agents.
    """

    # ------------------------------------------------------------------ #
    # Construction
    # ------------------------------------------------------------------ #
    def __init__(
        self,
        id_: str,
        description: str,
        tools: Sequence[Tool],
        check_last_step: bool = False,
    ) -> None:
        super().__init__(
            id_=id_,
            description=description,
            tools=tools,
            check_last_step=check_last_step,
        )
        self._execution_context: ExecutionContext | None = None

    # ------------------------------------------------------------------ #
    # Execution-context property (read-only from outside)
    # ------------------------------------------------------------------ #
    @property
    def execution_context(self) -> ExecutionContext:
        if self._execution_context is None:
            raise RuntimeError("ExecutionContext is not set.")
        return self._execution_context

    # ------------------------------------------------------------------ #
    # Database / scenario helpers
    # ------------------------------------------------------------------ #
    def get_db(self) -> ExecutionContext.DbConnector:
        return self.execution_context.db

    def get_scenario_id(self) -> str:
        return self.execution_context.scenario_id

    def get_run_id(self) -> str:
        return self.execution_context.run_id

    # ------------------------------------------------------------------ #
    # Public API
    # ------------------------------------------------------------------ #
    def execute(self, ctx: ExecutionContext, command: str) -> Step:
        """
        Run *command* inside *ctx* (simulation environment).

        The context is attached for the duration of the call and cleared
        afterwards, mimicking the Java `try/finally` behaviour.
        """
        self._execution_context = ctx
        try:
            return super().execute(command)
        finally:
            self._execution_context = None

    # ------------------------------------------------------------------ #
    # Tool life-cycle overrides (when LabAgent is itself used as a tool)
    # ------------------------------------------------------------------ #
    def init(self, agent: "Agent") -> None:  # type: ignore[override]
        if not isinstance(agent, ExecutorModule):
            raise ValueError("This tool can only be used by ReAct agents.")
        if not isinstance(agent.agent, LabAgent):
            raise ValueError("This tool can only be used by LabAgent instances.")
        super().init(agent)

    @property
    def agent(self) -> ExecutorModule:  # type: ignore[override]
        return super()._agent  # type: ignore[attr-defined]

    def get_lab_agent(self) -> "LabAgent":
        return self.agent.agent  # type: ignore[return-value]

    # ------------------------------------------------------------------ #
    # Tool invocation (when wrapped as a Tool)
    # ------------------------------------------------------------------ #
    def invoke(self, call: ToolCall) -> ToolCallResult:  # noqa: D401
        if not self.is_initialized():
            raise RuntimeError("Tool must be initialized.")

        question = self.get_string("question", call.arguments)
        if question is None:
            return ToolCallResult.from_call(
                call,
                'ERROR: You must provide a command to execute as "question" parameter.',
            )

        ctx = self.get_lab_agent().execution_context
        step = self.execute(ctx, question)

        if step.status == Status.ERROR:
            return ToolCallResult.from_call(call, f"ERROR: {step.observation}")

        return ToolCallResult.from_call(call, step.observation)

File: Nikhil/java/src/main/java/com/infosys/small/pnbc/LabAgent.java
package com.infosys.small.pnbc;

import java.util.List;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.infosys.small.core.Tool;
import com.infosys.small.core.ToolCall;
import com.infosys.small.core.ToolCallResult;
import com.infosys.small.pnbc.ExecutionContext.DbConnector;
import com.infosys.small.react.ExecutorModule;
import com.infosys.small.react.ToolableReactAgent;

import lombok.Getter;
import lombok.NonNull;

/**
 * This is an agent ({@link ToolableReactAgent}) that can interact with a
 * (simulated) backend system or orchestrate a (simulated) process. In addition,
 * it cna be used as a 6{@link Tool} by other agents.
 */
public class LabAgent extends ToolableReactAgent {

	/**
	 * Current execution context.
	 * 
	 * When set, it allows the agent to run in a simulated environment.
	 * 
	 * Notice this stays set only for the duration of
	 * {@link #execute(ExecutionContext, String)}
	 */
	@Getter
	private @NonNull ExecutionContext executionContext;

	/**
	 * Gets the connection to the database. This is valid only if execution context
	 * has been set.
	 */
	public DbConnector getDb() {
		return executionContext.getDb();
	}

	/**
	 * Gets the ID for the scenario we are currently running. This is valid only if
	 * execution context has been set.
	 */
	public String getScenarioId() {
		return executionContext.getScenarioId();
	}

	/**
	 * Gets the ID for current run. This is valid only if execution context has been
	 * set.
	 */
	public String getRunId() {
		return executionContext.getRunId();
	}

	/**
	 * @return The {@link LabAgent}that is using this tool, or null if it is not
	 *         running inside a LabAgent.
	 */
	public LabAgent getLabAgent() {
		if (getAgent() instanceof LabAgent) // Paranoid, should never be the case
			return (LabAgent) getAgent();
		if (getAgent() instanceof ExecutorModule) { // normally, it's in a ReactAgent inside a LabAgent
			ExecutorModule executor = (ExecutorModule) getAgent();
			if (executor.getAgent() instanceof LabAgent)
				return (LabAgent) executor.getAgent();
		}
		return null;
	}

	public LabAgent(@NonNull String id, @NonNull String description, @NonNull List<? extends Tool> tools) {
		this(id, description, tools, false);
	}

	public LabAgent(@NonNull String id, @NonNull String description, @NonNull List<? extends Tool> tools,
			boolean checkLastStep) {
		super(id, description, tools, checkLastStep);
	}

	/**
	 * Executes one command but it sets execution context first. This allows for
	 * having the agent running in a test (simulated) environment.
	 */
	public Step execute(ExecutionContext ctx, String command) throws JsonProcessingException {
		executionContext = ctx;
		return execute(command);
	}

	/**
	 * When invoked as a tool; it uses current execution context.
	 */
	@Override
	public ToolCallResult invoke(@NonNull ToolCall call) throws Exception {
		if (!isInitialized())
			throw new IllegalStateException("Tool must be initialized.");

		String question = getString("question", call.getArguments());
		if (question == null)
			return new ToolCallResult(call, "ERROR: You must provide a command to execute as \"question\" parameter.");

		ExecutionContext ctx = getLabAgent().getExecutionContext(); // Get execution context from caller
		Step result = execute(ctx, question);
		switch (result.status) {
		case ERROR:
			return new ToolCallResult(call, "ERROR: " + result.observation);
		default:
			return new ToolCallResult(call, result.observation);
		}
	}
}
